### DISCUSSION TITLE
Issue compiling project on HPC

### DISCUSSION THREAD

**Discussion by ajsummers** [idaholab/moose#16802] [Q&A Getting Started] [answered]
I previously posted an installation problem here: https://github.com/idaholab/moose/discussions/16711

Following from this, I have run into an error while running make for the first time on a project:
```
Compiling C++ (in opt mode) /home/ajs0201/projects/moose/framework/build/unity_src/outputs_Unity.C...
/usr/bin/ld: BFD version 2.20.51.0.2-5.48.el6_10.1 20100205 internal error, aborting at reloc.c line 443 in bfd_get_reloc_size

/usr/bin/ld: Please report this bug.

collect2: error: ld returned 1 exit status
make: *** [/home/ajs0201/projects/moose/modules/module_loader/lib/libmodule_loader_with-opt.la] Error 1
make: *** Waiting for unfinished jobs....
/usr/bin/ld: warning: libgfortran.so.4, needed by /home/ajs0201/miniconda3/envs/moose/lib/libpetsc.so, not found (try using -rpath or -rpath-link)
```

The operating system of the HPC is CentOS 6.10.
I have the following modules loaded into the environment:

- gcc/8.1.0
- mpich/3.2.1
- cmake/gcc/3.12.3
- mpc/1.0.3

Thanks again in advance for any help!

**Comment by ngrilli** [comment]
Dear @ajsummers 

Try the manual installation instructions:
https://mooseframework.inl.gov/getting_started/installation/manual_installation_gcc.html

It works better for us on CentOS.

Best Regards,
Nicolò Grilli
National University of Singapore

**Comment by milljm** [comment, marked as answer]
Going over your old post and this one, it looks like you did both a Conda and a Manual Install:

_I started off using the standard Linux installation method. Installed Miniconda and the Moose environment, then cloned in the repository. After this, I followed the GCC/MPICH manual installation method._

The error in question is from using Conda's moose-petsc, in conflict with compilers provided by your system:

```
    gcc/8.1.0
    mpich/3.2.1
    cmake/gcc/3.12.3
    mpc/1.0.3
```

These will conflict, as moose-petsc (10 days ago) was built with GCC 7.3.0 (Conda's compilers at the time: libgfortran.so.4). Today if you were to start over, you would see that moose-petsc is now using Conda's latest GCC compilers. -but that is neither here nor there. However, there would still be a conflict...

When deciding which set of instructions to follow, it is important to follow only one set. This is because:

- When using Conda (and specifically the moose packages moose-libmesh, or moose-petsc), there is no reason to build PETSc (or libMesh).
- When following the Manual Installation method, you have to build _everything_ manually (possibly gcc and mpich. Certainly PETSc, libMesh).

Because it looks like you were able to successfully build PETSc back in the old thread, I would recommend uninstalling (or simply not using) the moose environment you installed with Conda. If you followed the Conda instructions to the letter, that would mean you should do:

```
conda deactivate
conda env remove -n moose
```

Next, we should re-build libMesh, MOOSE and your application cleanly.

Close your terminals (for good measure). Re-open them, and clean moose and your application **(note! be sure to commit your changes or this will delete them!)**

```
cd /home/ajs0201/projects/moose
git clean -xfd
cd yourapp #<-- where ever this may be
git clean -xfd     # Note: again, be sure to commit any changes. This command deletes un-tracked files.
```

The above will purge any stale libraries lingering around, possibly still built with Conda's compilers.
Next, re-build libMesh, MOOSE and your app using your system compilers:

```
module purge
module load gcc/8.1.0 mpich/3.2.1 cmake/gcc/3.12.3 mpc/1.0.3
cd /home/ajs0201/projects/moose/scripts
./update_and_rebuild_libmesh.sh
cd yourapp
make
```

**Reply by ajsummers** [reply]
Thank you for your response. I will try this and update the thread. Should I recompile petsc as well?

**Reply by ajsummers** [reply]
I followed the directions in this post and got the following message while performing `make` in the test directory:
```
There appears to be a libMesh update in progress for
the branch of MOOSE you are operating on (only the
master branch contains publicly available moose-libmesh
packages).

You must provide your own libMesh either by:

    1. Uninstall moose-libmesh and build it manually via
       moose/scripts/update_and_rebuild_libmesh.sh
or
    2. Use `conda build` to build your own moose-libmesh
       package.
```

I've decided to remove Anaconda entirely and restart the GCC/MPICH installation from scratch.

**Reply by milljm** [reply]
_should i recompile petsc_
No need to rebuild PETSc, as I think you have that built correctly.

It may be that you have conda-build available. Honestly that little check is causing more issues than it solves... Is there a way you can create a new vanilla conda environment, so as not to have conda-build available?

```
conda create -n bare -q -y
conda activate bare
```

Should be pretty blank. Unless someone added some defaults that get installed with new environment created...

Sorry for the delayed response!

**Reply by milljm** [reply]
Oh! I didn't see this little tidbit here:

_I've decided to remove Anaconda entirely and restart the GCC/MPICH installation from scratch._

That's fine, sorry I wasn't able to respond in a timely manner. If you do happen to run into that "There appears to be a libMesh update...", please try the above conda bare environment trick (however, without Anaconda even installed, I foresee no issues). If that doesn't work, I know exactly where that little hiccup is happening in MOOSE, and we can avoid it (line 8 or ~10 in moose/framework/moose.mk). You will see it calling a conda verify script in moose/scripts. That is our culprit.

It is normally only triggered if someone has conda-build installed in their conda environment (what we use to _build_ moose-libmesh). And such, it sometimes gets in the way of legitimate builders trying to manually build the libraries such as yourself.
